### PR TITLE
[#1498] Frontend: maintain file type label after filtering

### LINKED ISSUE #1498: Zoom inconsistent labels
image

image

When I untick `tests`, the commit containing both `tests` and `backend` remain, but the label for `tests` disappears.

Even if this is intended behavior, I find it rather unintuitive. What do you think?

**If possible, include the URL to your RepoSense report or log files (if any).**
https://dashboard-1462-pr-reposense-reposense.surge.sh/

### PR DESCRIPTION
**Pull request by jbkim1999**
Fixes #1498

## Proposed commit message

```
When a commit contains two or more file types and one of them is
filtered out, the commit could still be visible if it has other file
types that are still filtered in.

However, in such cases, the filtered-out file type label is no longer
visible for that commit. This could confuse the users by indicating that
the commit no longer consists of that filtered-out file type.

Let's retain all the file types of a commit by returning all the file
types if at least one of them is still filtered in and hence viewable to
the users.
```

### When All is checked (See Add edit action... commit)
image

### After html file is filtered out (previous behaviour)
image

Same commit is visible, but without the html label (which should still be existent)

### After html file is filtered out (fixed behaviour)
image

html label is now visible.

### CODE REVIEW

File: frontend/src/views/v-zoom.vue [hunk 1]
```diff
@@ -101,8 +101,9 @@
       .hash
         span {{ slice.hash.substr(0, 7) }}
       span.fileTypeLabel(
+        v-if="containsAtLeastOneSelected(Object.keys(slice.fileTypesAndContributionMap))"
```
Comment (yhtMinceraft1010X) [inline frontend/src/views/v-zoom.vue:104]: Try adding a comma here to address the failing CI.

Comment (jbkim1999) [reply frontend/src/views/v-zoom.vue]: Thank you, addressed it for my next commit!

**Review by yhtMinceraft1010X — APPROVED** [review]
LGTM

File: frontend/src/views/v-zoom.vue [hunk 2]
```diff
@@ -383,8 +384,14 @@ export default {
       window.encodeHash();
     },
 
-    filterSelectedFileTypes(fileTypes) {
-      return fileTypes.filter((fileType) => this.selectedFileTypes.includes(fileType));
+    containsAtLeastOneSelected(fileTypes) {
+      let includesSelectedFileTypes = false;
+      fileTypes.forEach((fileType) => {
+        if (this.selectedFileTypes.includes(fileType)) {
+          includesSelectedFileTypes = true;
```
Comment (Zhou-Jiahao-1998) [inline frontend/src/views/v-zoom.vue:391]: This is just my personal preference, but I think it is more neat to simple return `true` here and return `false` on **line 394**. That way we can terminate the function earlier and declare one less variable. However, the difference is minimal, so it's really just my personal view. Let me know what you think :)

**Review by Zhou-Jiahao-1998 — CHANGES_REQUESTED** [review]
@jbkim1999 Please refer to my comments. Thank you!

File: frontend/src/views/v-zoom.vue (hunk 2 above)
Comment (jbkim1999) [reply frontend/src/views/v-zoom.vue:391]: I think that's a very reasonable point and I'll make the fix soon. Thank you for the comment!

File: frontend/cypress/tests/zoomView/zoomView_selectFileTypes.js [hunk 3]
```diff
@@ -44,10 +44,7 @@ describe('check file types ', () => {
         .uncheck()
         .should('not.be.checked');
 
-    cy.get('.zoom__day > .commit-message > .fileTypeLabel')
-        .should('not.contain.text', 'java');
-
-    cy.get('.zoom__day > .commit-message > .fileTypeLabel')
-        .should('contain.text', 'gradle');
+    cy.get('#tab-zoom .fileTypes input[value="all"]')
+        .should('not.be.checked');
```
[the remark this answers is not in the document]
Comment (jbkim1999) [reply frontend/cypress/tests/zoomView/zoomView_selectFileTypes.js:48]: I see, I was a little afraid of hard coding but I do agree that some more test cases can be generated from this PR. Will make this change also. Thanks for the suggestion!

File: frontend/cypress/tests/zoomView/zoomView_selectFileTypes.js [hunk 4]
```diff
@@ -44,10 +44,67 @@ describe('check file types ', () => {
         .uncheck()
         .should('not.be.checked');
 
-    cy.get('.zoom__day > .commit-message > .fileTypeLabel')
-        .should('not.contain.text', 'java');
+    cy.get('#tab-zoom .fileTypes input[value="all"]')
+        .should('not.be.checked');
+  });
+
+  it('a commit of a specific file type should not be seen when that type is unchecked', () => {
```
Comment (Zhou-Jiahao-1998) [inline frontend/cypress/tests/zoomView/zoomView_selectFileTypes.js:51]: Maybe putting as `a commit of only one file type...` will be clearer for people who look at this for the first time?

**Review by Zhou-Jiahao-1998 — APPROVED** [review]
@jbkim1999 LGTM! Thanks for writing the cypress test cases. Let me know what you think about the mentioned point.

File: frontend/cypress/tests/zoomView/zoomView_selectFileTypes.js (hunk 4 above)
Comment (jbkim1999) [reply frontend/cypress/tests/zoomView/zoomView_selectFileTypes.js:51]: What do you think about `a commit should not be seen when all of its file types are unchecked`?

In this way the condition is not restricted to having only one file type.

**Review by dcshzj — APPROVED** [review]
LGTM!

**Pull request merged**